hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -880,6 +880,133 @@ class Emit[C](
+              val leftDataAddress = lPType.data.load(leftPVal.tcode[Long])
+              val rightDataAddress = rPType.data.load(rightPVal.tcode[Long])
+
+              val answerPArrayAddress = mb.genFieldThisRef[Long]()
+              val M = lShape(lPType.nDims - 2)
+              val N = rShape(rPType.nDims - 1)
+              val K = lShape(lPType.nDims - 1)
+
+              val LDA = M
+              val LDB = K
+              val LDC = M
+
+              cb.ifx((M cne 0L) && (N cne 0L) && (K cne 0L), {
+                cb.assign(answerPArrayAddress, outputPType.data.pType.allocate(region.code, (M * N).toI))
+                cb.append(outputPType.data.pType.stagedInitialize(answerPArrayAddress, (M * N).toI))
+                cb.append(lPType.elementType match {
+                  case PFloat32(_) =>
+                    Code.invokeScalaObject13[String, String, Int, Int, Int, Float, Long, Int, Long, Int, Float, Long, Int, Unit](BLAS.getClass, method = "sgemm",
+                      "N",
+                      "N",
+                      M.toI,
+                      N.toI,
+                      K.toI,
+                      1.0f,
+                      lPType.data.pType.firstElementOffset(leftDataAddress),
+                      LDA.toI,
+                      rPType.data.pType.firstElementOffset(rightDataAddress),
+                      LDB.toI,
+                      0.0f,
+                      outputPType.data.pType.firstElementOffset(answerPArrayAddress, (M * N).toI),
+                      LDC.toI
+                    )
+                  case PFloat64(_) =>
+                    Code.invokeScalaObject13[String, String, Int, Int, Int, Double, Long, Int, Long, Int, Double, Long, Int, Unit](BLAS.getClass, method = "dgemm",
+                      "N",
+                      "N",
+                      M.toI,
+                      N.toI,
+                      K.toI,
+                      1.0,
+                      lPType.data.pType.firstElementOffset(leftDataAddress),
+                      LDA.toI,
+                      rPType.data.pType.firstElementOffset(rightDataAddress),
+                      LDB.toI,
+                      0.0,
+                      outputPType.data.pType.firstElementOffset(answerPArrayAddress, (M * N).toI),
+                      LDC.toI
+                    )
+                })
+              },
+                {
+                  cb.assign(answerPArrayAddress, outputPType.data.pType.zeroes(mb, region.code, (M * N).toI))
+                }
+              )
+              val res = outputPType.construct(
+                outputPType.makeShapeBuilder(IndexedSeq(M, N)),
+                outputPType.makeColumnMajorStridesBuilder(IndexedSeq(M, N), mb),
+                answerPArrayAddress,
+                mb,
+                region.code)
+
+              EmitCode(Code._empty, false, PCode(pt, res)).toI(cb)
```

Comment:
I think this can be `IEmitCode.present(cb, PCode(pt, res))` to avoid the intermediate `EmitCode` object

Reply:
Agree that's better, just didn't know the function to do this. Thanks